### PR TITLE
`kctrl`: Flag to create namespace when adding new repo

### LINKED ISSUE #1001: `kctrl`: When adding a package repo, include option to create the namespace if not existing
**Describe the problem/challenge you have**
When I add a Carvel package repository to a cluster, I first need to create a dedicated namespace explicitly.

1. `kubectl create namespace kadras-packages`
2. `kctrl package repository add -r kadras-repo --url ghcr.io/arktonix/kadras-packages:0.5.1 -n kadras-packages`

**Describe the solution you'd like**
I'd like to skip the explicit creation of the namespace and have it created as part of the `kctrl package` command, similar to how the `helm` CLI works. For example, the functionality could be enabled via a `--create-namespace` flag.

`kctrl package repository add -r kadras-repo --url ghcr.io/arktonix/kadras-packages:0.5.1 -n kadras-packages --create-namespace`

**Anything else you would like to add:**
One of the reasons for this feature request is to make the onboarding experience even simpler when introducing people to Carvel packaging or when using a Carvel metapackage as the foundation of projects/tutorials relying on a Carvel-based platform.

With this feature, that would result in three operations rather than four:

1. deploy kapp-controller
2. add the package repository
3. install the package

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by ThomasVitale**
#### What this PR does / why we need it:

When adding a new package repository to a cluster, it's now possible to create the installation namespace automatically by specifying the "--create-namespace" flag.

#### Which issue(s) this PR fixes:

Fixes gh-1001

#### Does this PR introduce a user-facing change?

```release-note
Added a new flag `create-namespace` for creating the installation namespace when adding a new package repository to a cluster.
```

#### Additional Notes for your reviewer:

After a first review of this pull request to validate the behaviour of the new flag, I will submit a PR to update the documentation on the project website.

##### Review Checklist:

- [X] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [X] Relevant tests are added or updated
- [] Relevant docs in this repo added or updated
- [] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [X] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (praveenrewar) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go]: We should probably return the error if it's *not* of type [IsAlreadyExists](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsAlreadyExists) and ignore if it already exists.

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (100mik) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go]: nit: lets have a more explicit name
```suggestion
	corev1 "k8s.io/api/core/v1"
```

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go:162]: +1 to Praveens comment. 
We do something similar over [here](https://github.com/carvel-dev/kapp-controller/blob/develop/cli/pkg/kctrl/cmd/package/installed/create_or_update.go#L585-L595)

However, in the example we attempt an `update` if the service already exists. Where as here, we return the error if the error is `!IsAlreadyExists()`

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (100mik) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go:159]: Let's add a log here saying that we are going to create a namespace. Something like `Creating namespace 'asdf'`.
Let's use `o.statusUI.PrintMessagef()` for this. (Adds a nice little timestamp in front of the message)

I think the message we are printing when namespace already exists should use `statusUI` as well 🤔

**Review by 100mik — COMMENTED** [review]
Some thoughts!

In addition to these looks like there is a redundant log line being printed on line 187
(More specifically: `o.ui.PrintLinef("Waiting for package repository to be added")`)
That can be removed as a part of this PR but does not block the PR itself!

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (100mik) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go:159]: "here" being right before we make the call to create it

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (ThomasVitale) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go]: Good point! I changed it as suggested.

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (ThomasVitale) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go]: Updated

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (ThomasVitale) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go:159]: Done. I have used `statusUI` for both messages.

Comment (praveenrewar) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go:159]: @100mik @ThomasVitale Instead of saying `Creating namespace xyz` and `Namespace xyz already exists` every time (thinking about updated scenarios), should we say `Created namespace xyz` once when we create the namespace?

(No strong opinions)

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (praveenrewar) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go]:
```suggestion
			if !errors.IsAlreadyExists(err) {
			         return err
			} 
			o.statusUI.PrintMessagef("The namespace '%s' already exists", o.NamespaceFlags.Name)
```

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (ThomasVitale) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go:159]: I like that! I have updated the code to only print a message if the namespace gets created.

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (ThomasVitale) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go]: I have refactored that part to print only a message if the namespace is created as suggested in the previous comment.

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (praveenrewar) [inline cli/pkg/kctrl/cmd/package/repository/add_or_update.go]: Should we remove this line as well? Or you already did and forgot to push the change?

Comment (100mik) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go]: I think I am good with the changes, once this is resolved ^

File: cli/test/e2e/package_repository_test.go
Comment (100mik) [inline cli/test/e2e/package_repository_test.go:133]: Just one change, can we add deletion of these two namespaces to the [cleanup function](https://github.com/carvel-dev/kapp-controller/pull/1113/files#diff-09f0c0ae3712103375872b02f78bf3bc30f715f8386e471d6235b5cf54d3e349R25)?
Maybe move the variables storing the names as well so that we can do that, something like `existingRepoNamesapce` and `newRepoNamespace` maybe.

File: cli/pkg/kctrl/cmd/package/repository/add_or_update.go
Comment (ThomasVitale) [reply cli/pkg/kctrl/cmd/package/repository/add_or_update.go]: Thanks for catching this. You'd think after several commits I'd learn the lesson about signing off commits, but I keep missing that and had to rebase (my Git is configured with the `-S` parameter for signing, but the DCA sign-off is with `-s`). I guess while doing that, this message popped up again.

File: cli/test/e2e/package_repository_test.go
Comment (ThomasVitale) [reply cli/test/e2e/package_repository_test.go:133]: I'm trying doing this, but the tests are failing. I couldn't find a way to run the tests locally, so the troubleshooting is a bit slow.

Comment (ThomasVitale) [reply cli/test/e2e/package_repository_test.go:133]: What do you think about this solution? https://github.com/carvel-dev/kapp-controller/pull/1113/commits/82dac38b2d06fbb4fadca77bb169dfa60a992440

Comment (100mik) [reply cli/test/e2e/package_repository_test.go:133]: Sorry! Were we already deleting the namespaces in the Cleanup function? It looks like we are removing those bits in this commit. If that is the case we do not need this change

Comment (ThomasVitale) [reply cli/test/e2e/package_repository_test.go:133]: From what I understood, the cleanup function is removing the package and related namespace in all test cases except the 2 new tests added because I use different namespaces. So I added the explicit cleanup for those 2 new test cases to delete both package and namespace since they are not caught by the cleanup function.

Comment (100mik) [reply cli/test/e2e/package_repository_test.go:133]: Ah I see, I was wondering if we could be deleting the new namespaces in the cleanup function itself
Mainly, because a namespace left over from an interrupted test run could affect these tests.

The commit you linked seemed to be removing the deletion of those namespaces from the cleanup function which had me a bit confused.

Ideally : We would have the commands deleting the namespaces in the cleanup function itself, so that we make sure they do not exist before and after running the tests!
What do you think about that?

Comment (100mik) [reply cli/test/e2e/package_repository_test.go:133]: (To put it concisely, I think the RemoveClusterResource call for these two namespaces should be within the cleanup function)

**Review by 100mik — APPROVED** [review]
Changes look good to me! Thank you so much for contributing

@praveenrewar feel free to merge this away if it looks good to you 🚀

File: cli/test/e2e/package_repository_test.go
Comment (praveenrewar) [inline cli/test/e2e/package_repository_test.go]: We could probably just use env.Namespace instead of this. It would remove the overhead of creating/deleting one namespace.

Comment (praveenrewar) [reply cli/test/e2e/package_repository_test.go]: Sorry didn't take a look at the tests earlier 😅

Comment (ThomasVitale) [reply cli/test/e2e/package_repository_test.go]: I considered doing that, but then opted for creating it explicitly to make the test case resemble the current 2-steps way of adding a repo as opposed to the changes I made in the PR (1. create the namespace, 2. add a repo). But I can change it to use the general one, if you want :)

Comment (ThomasVitale) [reply cli/test/e2e/package_repository_test.go]: Updated the tests here to re-use env.Namespace: https://github.com/carvel-dev/kapp-controller/pull/1113/commits/893f9e7beb0e8faa1892be52c9a07e706e7e8002

**Review by praveenrewar — APPROVED** [review]
LGTM! Thank you so much for the contribution 🚀

**Comment by ThomasVitale** [comment]
@100mik @praveenrewar thanks a lot for your guidance and review

**Pull request merged**
